eZ/Publish/Core/Persistence/Cache/ContentHandler.php

```diff
@@ -276,8 +277,12 @@ public function deleteVersion($contentId, $versionNo)
     /**
      * {@inheritdoc}
      */
-    public function listVersions($contentId, $status = null, $limit = -1)
+    public function listVersions($contentId, $status = null, $limit = self::DEFAULT_LIST_VERSIONS_LIMIT)
     {
+        if ($limit !== self::DEFAULT_LIST_VERSIONS_LIMIT) {
+            return $this->persistenceHandler->contentHandler()->listVersions($contentId, $status, $limit);
```

Comment:
This disables cache also for [this operation](https://github.com/ezsystems/ezpublish-kernel/blob/v7.4.2/eZ/Publish/Core/Repository/ContentService.php#L1546-L1550), which can be bad. AdminUI is not affected as `$limit` is not exposed to the API (yet ;) ).

If it happens that cache doesn't return proper list of versions, with published ones, then the problem exists within cache invalidation. I looked briefly at tags that are invalidated and don't see an issue, so you need to dig in to this deeper to determine why we're getting invalid list of versions from cache. Hint: maybe filtering by status is not invalidated properly?

Reply:
Yes, that's correct. The issue is caused by incorrect cache invalidation.

Long story short: the cause of this bug is that result of `ContentHandler::loadVersions` is invalidated in `ContentHandler::deleteVersion` by invalidating on tag `content-_contentId_-version-_versionNo_`. 
Let's say, that we have 5 content versions with `versionNo`: 1,2,3,4,5.
At first, we remove version 5, so we invalidate on tag `content-_contentId_-version-5`, so `ez-content-_contentId_-version-list` gets invalidated, because it has been tagged using *all* version numbers before. Then, when we remove version 4, `ContentHandler::deleteVersion` tries to invalidate on `content-_contentId_-version-4`, but the fresh cache element under `ez-content-_contentId_-version-list` is *not* tagged by this version. It’s tagged *only* by `content-_contentId_-version-1` and `content-_contentId_-version-2` since that method has been executed with limit `2` internally by `ContentService::deleteVersion`. This is a little complicated, but I hope the explanation is somehow understandable.

Thanks for pointing out that this change indeed affects also publishing operation, therefore I'll slightly change the approach.

Reply:
Ok, I get it now. So we essentially cache result w/o taking into the account that it might be filtered. I think we need to approach caching this list differently. You should have `\eZ\Publish\Core\Persistence\Cache\AbstractHandler::getMultipleCacheItems` method introduced by @andrerom for bulk loading. Maybe it could help here as well?

Reply:
getMultipleCacheItems, or getMultipleCacheValues as it's called now in AbstractInMemoryHandler for handlers using that. Is for lookup by bulk id's, this is rather a list, abstract in AbstractInMemoryHandler->getListCacheValue.

@alongosz We can fix the key here to be unique depending on limit. But the question was if there was any point in caching edge cases and spending memory on that, specifically thinking  of the case that was originally fixed/worked around:
```
        $versionList = $this->persistenceHandler->contentHandler()->listVersions(
            $versionInfo->contentInfo->id,
            null,
            2
        );
```

WDYT?

Reply:
> @alongosz We can fix the key here to be unique depending on limit. But the question was if there was any point in caching edge cases and spending memory on that, specifically thinking of the case that was originally fixed/worked around:
> 
> ```
>         $versionList = $this->persistenceHandler->contentHandler()->listVersions(
>             $versionInfo->contentInfo->id,
>             null,
>             2
>         );
> ```
> WDYT?
> 
> But yes, we should make sure the publish case is cached. Maybe we should have constants for 25, 50 and 100, and cache usage of any of those but not strange cases in between.

This doesn't feel very clean to me. Someone adding new feature using other combination of arguments would surely miss such requirement. Or if that someone didn't miss it then he'd have to add another edge case to the cache implementation. This is very much not SOLID. The fact that we've initially, right now, missed publish version case kinda proves my point :-)

I'm accepting workarounds only if there's no better option at the moment (OT: working on such case right now ;) ).

However here I have another suggestion. If we don't want to waste memory on edge cases, I suggest to do the following:
1. Add to `listVersions` an optional argument/flag: `$disableCache = false`.
2. Make caching key unique for all combinations of arguments anyway (memoy won't be wasted if we don't use it when `$disableCache === true`)

You could argue of course that `$disableCache` belongs to the only one layer and in others won't be used. However when you look at layers as strategies, this is not uncommon that some options are used by specific strategies only. As a bonus you won't need to remember about it when adding another call to that method.

WDYT?

Reply:
Rather not, then let's just vary key by limit cleanly and stay clean.

As in: If something is cached or not is not something caller should need to know, and we shouldn't have to pollute our SPI and API with explicit logic around this.

Reply:
Thanks @alongosz & @andrerom for your great input on that. Please, review again. I did clean-up and applied what @andrerom mentioned in https://github.com/ezsystems/ezpublish-kernel/pull/2571#discussion_r267691629.

Reply:
Isn't using `Cache\ContentHandler` to get uncached result a bit strange in a first place? We have an option to inject `Legacy\Content\Handler` for such cases. Or maybe, for edge cases anyway, it would be more consistent and predictable to return `array_slice` of non-limited cache hit. We would avoid memory waste yet still hit the cache.

Reply:
@Nattfarinn that's correct. That's why finally limited results are cached as well. This is not widely used around the system, so I hope it's not gonna be a big deal in regards to memory consumption.